src/urllib3/contrib/pyopenssl.py

```diff
@@ -456,21 +457,33 @@ def wrap_socket(
     ):
         cnx = OpenSSL.SSL.Connection(self._ctx, sock)
 
-        if isinstance(server_hostname, str):
-            server_hostname = server_hostname.encode("utf-8")
-
-        if server_hostname is not None:
-            cnx.set_tlsext_host_name(server_hostname)
+        # If server_hostname is an IP, don't use it for SNI, per RFC6066 Section 3
+        if server_hostname and not util.is_ipaddress(server_hostname):
+            server_hostname_bytes = server_hostname
+            if isinstance(server_hostname, str):
+                server_hostname_bytes = server_hostname_bytes.encode("utf-8")
+            cnx.set_tlsext_host_name(server_hostname_bytes)
 
         cnx.set_connect_state()
 
         while True:
             try:
                 cnx.do_handshake()
+                if self.check_hostname:
+                    if util.is_ipaddress(server_hostname):
```

Comment:
What are we gaining by using `service-identity` package over our stricter verification of IP addresses from the standard library which only checks equality?

Since we're keeping around `match_hostname()` to disable commonName checking on Python 3.6 can we use `match_hostname()` here instead of bringing on a new dependency?

Reply:
Right, it would be much better to use our own code here, thanks!